filip/utils/cleanup.py

```diff
@@ -0,0 +1,157 @@
+"""
+Functions to clean up a tenant within a fiware based platform.
+
+created Oct 08, 2021
+
+@author Thomas Storek
+"""
+import functools
+from requests import RequestException
+from filip.models import FiwareHeader
+from filip.clients.ngsi_v2 import ContextBrokerClient, IoTAClient, QuantumLeapClient
+
+
+def clear_cb(url: str, fiware_header: FiwareHeader):
+    """
+    Function deletes all entities, registrations and subscriptions for a
+    given fiware header
+
+    Note:
+        Always clear the devices first because the IoT-Agent will otherwise
+        through errors if it cannot find its registration anymore.
+
+    Args:
+        url: Url of the context broker service
+        fiware_header: header of the tenant
+
+    Returns:
+        None
+    """
+    # create client
+    client = ContextBrokerClient(url=url, fiware_header=fiware_header)
+
+    # clear entities
+    entities = client.get_entity_list()
+    if entities:
+        client.update(entities=entities, action_type='delete')
+
+    # clear subscriptions
+    for sub in client.get_subscription_list():
+        client.delete_subscription(subscription_id=sub.id)
+
+    # clear registrations
+    for reg in client.get_registration_list():
+        client.delete_registration(registration_id=reg.id)
+
+
+def clear_iota(url: str, fiware_header: FiwareHeader):
+    """
+    Function deletes all device groups and devices for a
+    given fiware header
+
+
+
+    Args:
+        url: Url of the context broker service
+        fiware_header: header of the tenant
+
+    Returns:
+        None
+    """
+    # create client
+    client = IoTAClient(url=url, fiware_header=fiware_header)
+
+    # clear groups
+    for group in client.get_group_list():
+        client.delete_group(resource=group.resource,
+                            apikey=group.apikey)
+
+    # clear registrations
+    for device in client.get_device_list():
+        client.delete_device(device_id=device.device_id)
+
+
+def clear_ql(url: str, fiware_header: FiwareHeader):
+    """
+    Function deletes all data for a given fiware header
+    Args:
+        url: Url of the quantumleap service
+        fiware_header: header of the tenant
+
+    Returns:
+        None
+    """
+    # create client
+    client = QuantumLeapClient(url=url, fiware_header=fiware_header)
+
+    # clear data
+    try:
+        for entity in client.get_entities():
+            client.delete_entity(entity_id=entity.entityId,
+                                 entity_type=entity.entityType)
+    except RequestException as err:
+        if err.response.status_code == 404:
+            try:
+                err.response.json()['error'] == 'Not Found'
+            except KeyError:
```

Comment:
Shouldn't be also an error raised if the 'error' entry is not 'not found' ?

Reply:
@RedingJeff actually, with clear we want that error message because it tells us that not entries are in the database

Reply:
@tstorek OK, but does the == statement here have any effect? Doesnt the code only tries if it can access the "error" key of the json() without error?

Reply:
@RedingJeff good point 👍